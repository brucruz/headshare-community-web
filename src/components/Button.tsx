import { ButtonHTMLAttributes } from 'react';
import { ButtonContainer } from '../styles/components/Button';

export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  priority?: 'primary' | 'secondary' | 'tertiary';
  text: string | any[];
  size?: 'medium' | 'small';
  icon?: any;
  stretch?: boolean;
  isLoading?: boolean;
}

const Button: React.FC<ButtonProps> = ({
  priority = 'primary',
  text,
  stretch = false,
  disabled = false,
  size = 'medium',
  icon,
  isLoading = false,
  ...rest
}) => (
  <ButtonContainer
    disabled={disabled}
    priority={priority}
    stretch={stretch}
    size={size}
    isLoading={isLoading}
    {...rest}
  >
    {icon && icon}
    {size === 'small' ? <p>{text}</p> : <h4>{text}</h4>}
  </ButtonContainer>
);

export default Button;
