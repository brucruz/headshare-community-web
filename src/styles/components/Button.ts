import styled, { css } from 'styled-components';

interface ButtonProps {
  priority: 'primary' | 'secondary' | 'tertiary';
  stretch?: boolean;
  size: 'medium' | 'small';
  isLoading?: boolean;
}

export const ButtonContainer = styled.button<ButtonProps>`
  height: 38px;
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 19px;
  font-weight: 600;
  cursor: pointer;

  ${props =>
    props.disabled &&
    css`
      opacity: 0.6;
      cursor: not-allowed;
    `}

  ${props =>
    props.stretch &&
    css`
      width: 100%;
    `}

  h4 {
    margin: 8px 18px;
    font-weight: 600;
  }

  p {
    margin: 3px 10px;
  }

  ${props =>
    props.priority === 'primary' &&
    css`
      background-color: #e74f4f;
      color: #fafafa;
    `}

  ${props =>
    props.priority === 'secondary' &&
    css`
      border: 2px solid #e74f4f;
      color: #e74f4f;
    `}

  ${props =>
    props.priority === 'tertiary' &&
    css`
      background-color: #e3e3e3;
      color: #363636;
    `}

  ${props =>
    props.size === 'small' &&
    css`
      height: 35px;
      border-radius: 17.5px;
    `}

  ${props =>
    !props.disabled &&
    css`
      &:hover {
        box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25);

        opacity: 0.9;
      }
    `}

  ${props =>
    props.isLoading &&
    css`
      cursor: auto;

      &:hover {
        box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25);

        opacity: 0.9;
      }
    `}
`;
