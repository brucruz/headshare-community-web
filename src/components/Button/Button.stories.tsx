/* eslint-disable import/no-extraneous-dependencies */
import { Story, Meta } from '@storybook/react';
// eslint-disable-next-line import/no-unresolved

import Button, { ButtonProps } from './index';

export default {
  title: 'Headshare/Button',
  component: Button,
} as Meta;

const Template: Story<ButtonProps> = args => <Button {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  priority: 'primary',
  text: 'Click here',
  size: 'medium',
  stretch: false,
  isLoading: false,
};

export const Secondary = Template.bind({});
Secondary.args = {
  priority: 'secondary',
  text: 'Click here',
  size: 'medium',
  stretch: false,
  isLoading: false,
};

export const Tertiary = Template.bind({});
Tertiary.args = {
  priority: 'tertiary',
  text: 'Click here',
  size: 'medium',
  stretch: false,
  isLoading: false,
};
