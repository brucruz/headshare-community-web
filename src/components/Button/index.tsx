import { ButtonHTMLAttributes } from 'react';
import { ButtonContainer } from '../../styles/components/Button';
import { LoadingAnimation } from '../LoadingAnimation';

export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  priority?: 'primary' | 'secondary' | 'tertiary';
  text: string | any[];
  size?: 'medium' | 'small';
  icon?: any;
  stretch?: boolean;
  isLoading?: boolean;
}

const Button: React.FC<ButtonProps> = ({
  priority = 'primary',
  text,
  stretch = false,
  disabled = false,
  size = 'medium',
  icon,
  isLoading = false,
  ...rest
}) => (
  <ButtonContainer
    disabled={disabled}
    priority={priority}
    stretch={stretch}
    size={size}
    isLoading={isLoading}
    {...rest}
  >
    {icon && icon}
    {size === 'small' ? (
      <p>
        {isLoading && (
          <LoadingAnimation
            color={priority === 'primary' ? 'secondary' : 'primary'}
            application="button"
          />
        )}
        {text}
      </p>
    ) : (
      <h4>
        {isLoading && (
          <LoadingAnimation
            color={priority === 'primary' ? 'secondary' : 'primary'}
            application="button"
          />
        )}
        {text}
      </h4>
    )}
  </ButtonContainer>
);

export default Button;
