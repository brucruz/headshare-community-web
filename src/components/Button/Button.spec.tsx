/* eslint-disable import/no-extraneous-dependencies */
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
// import initStoryshots from '@storybook/addon-storyshots';

// ðŸ‘‡ Imports a specific story for the test
import { Primary } from './Button.stories';

// initStoryshots();

describe('<Button />', () => {
  it('renders in the primary state', () => {
    render(<Primary {...Primary.args} />);
    expect(screen.getByRole('button')).toHaveTextContent('Click here');
    expect(screen.getByRole('button')).toMatchSnapshot();
  });
});
