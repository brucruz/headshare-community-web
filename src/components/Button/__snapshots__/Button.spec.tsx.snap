// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders in the primary state 1`] = `
<button
  class="Button__ButtonContainer-sc-3gfsk5-0 jekWvy"
>
  <h4>
    Click here
  </h4>
</button>
`;
